Evaluate ``keccak256(a, const)``, when the value at memory location ``a`` is known at compile time and ``const <= 32``

hrkrshnn opened the pull request:
Implements part of https://github.com/ethereum/solidity/issues/11018.

Evaluating multiple words will need some symbolic reasoning, which we don't currently have.

hrkrshnn committed:
Evaluate `keccak(a, const)` if value at memory location `a` is known

Here the value of constant can be at most 32.

hrkrshnn commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -46,13 +51,31 @@ void LoadResolver::visit(Expression& _e)
 				break;
 			}
+
+		if (isKeccak(funCall->functionName.name))
```
Should we add some checks for `runs` parameter here?

chriseth replied:
Yes, please, the big constants are rather expensive.

hrkrshnn replied:
Added a check based on optimize-runs to see if the rule should be applied.

ekpyron commented on test/libyul/yulOptimizerTests/loadResolver/keccak_fail2.yul:
```diff
@@ -0,0 +1,15 @@
+{
+    mstore(0, 10)
+    // Although, this can be evaluated, we're skipping it.
```
That's not what the expectation looks like right now, though, is it?

hrkrshnn replied:
Oops. Fixed it. It was a copy and paste issue.

ekpyron commented:
> Evaluating multiple words will need some symbolic reasoning, which we don't currently have.

Can you elaborate on that? While evaluating only for a length of 32 is a fine first step in any case, I'm not sure I see what'd prevent us to try evaluating arbitrary known sizes...

ekpyron commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +99,40 @@ void LoadResolver::tryResolve(
+			_e = Literal{
+				locationOf(_e),
+				LiteralKind::Number,
+				YulString{u256(keccak256(toBigEndian(*memoryContent))).str()},
```
Is there any particular reason why you don't just truncate `toBigEndian(*memoryContent)` to a size of `memoryOffset` for `memoryOffset < 32`?

hrkrshnn replied:
Good idea. I'll implement it!

ekpyron commented:
> > Evaluating multiple words will need some symbolic reasoning, which we don't currently have.
> 
> Can you elaborate on that? While evaluating only for a length of 32 is a fine first step in any case, I'm not sure I see what'd prevent us to try evaluating arbitrary known sizes...

Ah, nevermind, I see what you mean... if we know the literal value of the memory offset it'd be fine, but not if it's some variable and we only know the contents...

hrkrshnn commented:
@ekpyron The main problem is that `m_memory` from DataFlowAnalyzer currently maps `YulString -> YulString`.

So we can only evaluate `keccak(a, 32)` if `mstore(a, value)` exists. (Even `let b := a; keccak(b, 32)` won't be evaluated, although further optimizations would simplify it and make the evalutation possible.)

The problem with evaluating, say `keccak(a, 64)` is to find the value at the memory location `add(a, 32)`. We could probably use `KnowledgeBase` to compute what variable is equivalent to `add(a, 32)` and then find its value from `m_memory`. Basically we need some symbolic reasoning here.

ekpyron commented:
I'm wondering if it's worthwhile to optimize for the cases in which the actual value of the offset is known to be zero.
Just because mapping index access uses ``keccak256(0, 0x40)``...

ekpyron commented:
In any case, only length 32 is, as I said, good enough for a first step in any case.

hrkrshnn commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +102,44 @@ void LoadResolver::tryResolve(
+		if (memoryContent && memoryOffset <= 32)
+		{
+			bytes contentAsBytes = toBigEndian(*memoryContent);
+			bytes slicedBytes = contentAsBytes | ranges::views::take(unsigned(*memoryOffset)) | ranges::to<vector>;
```
```suggestion
			bytes slicedBytes = toBigEndian(*memoryContent) | ranges::views::take(unsigned(*memoryOffset)) | ranges::to<vector>;
```

doesn't get type checked. Probably because it's a range of an rvalue. I wonder if there is a cleaner way to write this.

ekpyron replied:
Hm... ``ranges::views::counted(begin(toBigEndian(*memoryContent)), *memoryOffset) | ranges::to<bytes>`` *should* work, but is also not nice...

chriseth replied:
Shouldn't this be `32 - *memoryOffset` (because it is actually "length')?

chriseth commented on Changelog.md:
```diff
@@ -7,6 +7,7 @@ Compiler Features:
  * AST: Export NatSpec comments above each statement as their documentation.
  * Inline Assembly: Do not warn anymore about variables or functions being shadowed by EVM opcodes.
  * Optimizer: Simple inlining when jumping to small blocks that jump again after a few side-effect free opcodes.
+ * Yul Optimizer: Evaluate ``keccak256(a, c)``, when the value at memory location ``a`` is known at compile time and ``c`` is a constant ``<= 32``.
```
Can this optimization opportunity be used for the old code generator?

hrkrshnn replied:
Not sure.

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +102,44 @@ void LoadResolver::tryResolve(
+	};
+
+	yulAssert(_arguments.size() == 2, "");
+	Identifier const* memoryKey =  std::get_if<Identifier>(&_arguments.at(0));
```
```suggestion
	Identifier const* memoryKey = std::get_if<Identifier>(&_arguments.at(0));
```

chriseth commented:
The load resolver is getting rather complicated. I see that it does not make sense to make this its own run for performance reasons, but maybe it would make sense to split it into modules, where each module is queried to simplify a builtin or other function call?

bshastry commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +102,44 @@ void LoadResolver::tryResolve(
+	Identifier const* memoryKey =  std::get_if<Identifier>(&_arguments.at(0));
+	Identifier const* offset = std::get_if<Identifier>(&_arguments.at(1));
+
+	if (!memoryKey || !offset) return;
```
Not sure if our coding guidelines expect a return on the next line

ekpyron replied:
I think I never saw it on the same line in any case...

hrkrshnn replied:
Fixed.

hrkrshnn committed:
Optimizer context has parameter expectedExecutionsPerDeployment

hrkrshnn committed:
Added command line tests for keccak optimization with low runs

The value of keccak256(0, 32) should not be replaced by the big constant

hrkrshnn commented:
> The load resolver is getting rather complicated. I see that it does not make sense to make this its own run for performance reasons, but maybe it would make sense to split it into modules, where each module is queried to simplify a builtin or other function call?

Right now, there is only a single function `tryEvaluateKeccak`. If we split it into classes, we'll have to pass references to `m_memory`; not sure if this is much better. Is that what you were referring to?

ekpyron commented on libyul/AssemblyStack.cpp:
```diff
@@ -185,7 +185,9 @@ void AssemblyStack::optimize(Object& _object, bool _isCreation)
-		m_optimiserSettings.yulOptimiserSteps
+		m_optimiserSettings.yulOptimiserSteps,
+		{},
+		m_optimiserSettings.expectedExecutionsPerDeployment
```
I wonder whether the optimizer context shouldn't just get the full optimizer settings... but can be done in the future, no need to change the PR because of it.
But more importantly: this should be
```suggestion
		_isCreation ? nullopt : m_optimiserSettings.expectedExecutionsPerDeployment
```
shouldn't it?

ekpyron commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+		m_expectedExecutionsPerDeployment.has_value() &&
+		bigint(*m_expectedExecutionsPerDeployment) * runTimeGasSavings < deployTimeGasLoss
+	)
+		return;
```
We should really have backtracking... I wonder how likely this would be to become cheaper due to the possibility to simplify more stuff based on knowing the value later on...

chriseth replied:
Do we at least combine two subsequent keccak operations of the same memory location?

hrkrshnn replied:
> Do we at least combine two subsequent keccak operations of the same memory location?

Yes. Even CSE would probably do that.

bshastry commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+	// instead of
+	// [PUSH1 0 PUSH1 32 KECCAK256]: 5 bytes
+	// Increase of 28 bytes.
+	// Current increase: 28 * 200: 56000
```
```suggestion
	// Current increase: 28 * 200 = 56000
```

hrkrshnn replied:
Ah. There is also a typo. It should be `5600`. :(

ekpyron commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+	)
+		return;
+
+	auto valueOfIdentifier = [&](YulString const& _name) -> optional<u256>
```
I myself like using local lambdas for stuff like this and tend to be annoyed when reviewers suggest to change it... but for this one here I wonder whether it wouldn't even make sense to move it back to the DataFlowAnalyzer as protected member function...

hrkrshnn replied:
Sure, I'll make it a member function of `DataFlowAnalyzer`.

bshastry commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+	// [PUSH1 0 PUSH1 32 KECCAK256]: 5 bytes
+	// Increase of 28 bytes.
+	// Current increase: 28 * 200: 56000
+	unsigned deployTimeGasLoss = 28 * GasCosts::createDataGas;
```
```suggestion
	unsigned constexpr deployTimeGasLoss = 28 * GasCosts::createDataGas;
```

bshastry commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+{
+	// Runtime saving: [PUSH32 32-byte-value] instead of [PUSH1 0 PUSH1 32 KECCAK256]
+	// Saves one KECCAK256 + one PUSH. Currently it is 30 + 6 + 3 = 39
+	unsigned runTimeGasSavings = GasCosts::keccak256Gas + GasCosts::keccak256WordGas + GasCosts::tier2Gas;
```
```suggestion
	unsigned constexpr runTimeGasSavings = GasCosts::keccak256Gas + GasCosts::keccak256WordGas + GasCosts::tier2Gas;
```

bshastry commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+	// We skip if the net savings are not positive
+	if (
+		m_expectedExecutionsPerDeployment.has_value() &&
+		bigint(*m_expectedExecutionsPerDeployment) * runTimeGasSavings < deployTimeGasLoss
```
Why is `bigint` conversion needed?

bshastry replied:
Just wondering if this could be made simpler by the following check

```
if (*m_expectedExecutionsPerDeployment < 435)
  return;
```

since gas saving/loss are const expressions.

hrkrshnn replied:
> Why is bigint conversion needed?

Was worried about overflow, since some contract writers seems to set very high `m_expectedExecutionsPerDeployment`.

> `if (*m_expectedExecutionsPerDeployment < 435)`

I did not want to hardcode the value since gas prices may change. Also, the RHS should be around 148.

bshastry replied:
I suppose the `runs` parameter cannot exceed maximum of `size_t` datatype, right? That is the data type of the runs parameter in optimiser settings struct.

hrkrshnn replied:
> I suppose the runs parameter cannot exceed maximum of size_t datatype.

Yes. But when `runs` is `numeric_limits<size_t>::max()`, the multiplication would overflow or get into undefined behaviour. Doing it in `bigint` solves this issue. Alternatively, we could simply add an extra condition `size_t < 10000 && (....)`, if you think the `bigint` conversion is not nice. We seem to use `bigint` rather generously in different parts, though.

bshastry reviewed (commented):
First review round. Will follow this up with more comments shortly.

bshastry commented on test/libsolidity/semanticTests/constructor/arrays_in_constructors.sol:
```diff
@@ -26,6 +26,6 @@ contract Creator {
 // ----
 // f(uint256,address[]): 7, 0x40, 10, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 -> 7, 8
-// gas irOptimized: 472538
+// gas irOptimized: 478315
```
Any idea why an increase by close to `5k` here?

hrkrshnn replied:
The difference is evaluating `keccak` inside a constructor. Even though currently, `keccak` doesn't get evaluated for deploy code, this example seems to be an edge case. Here is the diff: https://gist.github.com/hrkrshnn/61eacc72681811ab5d3d992b2fcab8ca#file-keccak-diff

Let me see if this can be fixed.

hrkrshnn replied:
I made an issue about what is causing this: https://github.com/ethereum/solidity/issues/11118. Would be rather difficult to improve this now.

ekpyron commented on libyul/optimiser/OptimiserStep.h:
```diff
@@ -37,6 +37,7 @@ struct OptimiserStepContext
 	Dialect const& dialect;
 	NameDispenser& dispenser;
 	std::set<YulString> const& reservedIdentifiers;
+	std::optional<size_t> expectedExecutionsPerDeployment;
```
Is ``nullopt`` meant to be used for creation code? If so maybe that should be documented?

hrkrshnn replied:
Currently, the wasm code transform didn't have access to `expectedExecutionsPerDeployment`, so I made it `nullopt` and the `keccak` would get evaluated.

Similarly, also for `YulOptimizerTest`. Although, for this, we could just replace `nullopt` by `200`.

bshastry replied:
For the fuzzer to test the keccak rule in this PR, the check on expected runs may either be disabled (current state i.e., in fuzzer tests rule takes effect always) or set to a value that passes that check i.e., a runs that provides gas savings.

Since the fuzzer relies on YulOptimizerTest, please either set the runs to `nullopt` or some value that is large enough for there to be gas savings.

ekpyron commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+	unsigned deployTimeGasLoss = 28 * GasCosts::createDataGas;
+	// We skip if the net savings are not positive
+	if (
+		m_expectedExecutionsPerDeployment.has_value() &&
```
So given this I guess ``!m_expectedExecutionsPerDeployment`` is really supposed to mean creation code and for creation code it always pays off? Might also be nice to have a comment about that...

hrkrshnn replied:
I didn't really think much about creation code. Why does it always pay off for creation? Even for `create` and `create2` inside contracts?

ekpyron replied:
Actually, I was talking nonsense. The opposite is true - for creation code you can always assume ``m_expectedExecutionsPerDeployment`` being 1 (since, well, the deploy code is executed once per deployment)... so the condition below would always be true, so if ``m_expectedExecutionsPerDeployment == nullopt`` were supposed to mean creation code, then the condition would probably have to be ``!m_expectedExecutionsPerDeployment.has_value() || ...``....

hrkrshnn replied:
Why does it have to be `1`? Are you saying that `keccak` inside constructors should not be evaluated?

ekpyron replied:
Well, by definition, you execute the deploy code once per deployment, so the "expected executions per deployment" for deploy code is one.
I guess you're wondering about the case in which a contract keeps creating instances of another contract, so the deploy code is run several times in that sense? That's not really captured by "expectedExecutionsPerDeployment" I guess...

ekpyron replied:
By the way: you'd probably get around https://github.com/ethereum/solidity/pull/11026#discussion_r594225114 if you didn't evaluate during constructors.

ekpyron replied:
Independently, the gas costs would have to be calculated differently for deploy code.

ekpyron replied:
(e.g. in the inliner I added ``u256 GasMeter::dataGas(uint64_t _length, bool _inCreation, langutil::EVMVersion _evmVersion)`` to calculate the data gas cost of `_length` bytes depending on whether it's creation code or not - runtime execution cost is the same)

ekpyron reviewed (dismissed):
Some comments - the main thing that I think should be done is to clarify how it deals with creation code.

bshastry reviewed (commented):
Small review round.

hrkrshnn commented:
Ready for review again. The following changes since last time:

1. Moved `valueOfIdentifier` into `DataFlowAnalyzer`.
2. The creation code calculation is changed. But https://github.com/ethereum/solidity/pull/11026#discussion_r594225114 is deeper than that. Need to solve https://github.com/ethereum/solidity/issues/11118 for a general fix.

hrkrshnn commented:
Rebased on develop.

bshastry commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+	// instead of
+	// [PUSH1 0 PUSH1 32 KECCAK256]: 5 bytes
+	// Increase of 28 bytes.
+	// Current increase: 28 * 200 = 56000, when it's not deploy code.
```
```suggestion
	// Current increase: 28 * 200 = 5600, when it's not deploy code.
```
```suggestion
	// Current increase: 28 * 200 = 56000, when it's not deploy code.
```

hrkrshnn replied:
Fixed.

bshastry commented on test/libyul/YulOptimizerTestCommon.cpp:
```diff
@@ -311,7 +311,7 @@ YulOptimizerTestCommon::YulOptimizerTestCommon(
 		{"fullSuite", [&]() {
 			GasMeter meter(dynamic_cast<EVMDialect const&>(*m_dialect), false, 200);
-			OptimiserSuite::run(*m_dialect, &meter, *m_object, true, solidity::frontend::OptimiserSettings::DefaultYulOptimiserSteps);
+			OptimiserSuite::run(*m_dialect, &meter, *m_object, true, solidity::frontend::OptimiserSettings::DefaultYulOptimiserSteps, 200);
```
Does `runs = 200` pass this check i.e., enable the optimisation in this PR?

https://github.com/ethereum/solidity/pull/11026/files#diff-f9c9dce78307a3b319b2c54b0baf3911125ccd830ebfca1a45ce44eab92a07fdR133

bshastry replied:
Looks like it, just cross-checking.

hrkrshnn commented:
<details><summary>Click for a table of gas differences</summary>

| file name | IR-Opti | Legacy-Opti | Legacy|
|-----------|---------|-------------|-------|
| abiEncoderV2/abi_encode_v2.sol |     -0.035% |      0.000% |      0.000% |
| abiEncoderV2/storage_array_encoding.sol |     -0.023% |      0.000% |      0.000% |
| array/copying/array_copy_calldata_storage.sol |     -0.007% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_dynamic_dynamic.sol |     -0.014% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_static_dynamic.sol |     -0.045% |      0.000% |      0.000% |
| array/copying/array_copy_storage_to_memory_nested.sol |     -0.070% |      0.000% |      0.000% |
| array/copying/array_nested_calldata_to_storage.sol |     -0.024% |      0.000% |      0.000% |
| array/copying/copy_function_storage_array.sol |     -0.029% |      0.000% |      0.000% |
| array/delete/delete_storage_array_packed.sol |     -0.038% |      0.000% |      0.000% |
| array/push/array_push.sol |     -0.077% |      0.000% |      0.000% |
| constructor/arrays_in_constructors.sol |      1.222% |      0.000% |      0.000% |
| various/destructuring_assignment.sol |     -0.016% |      0.000% |      0.000% |
| viaYul/array_storage_index_zeroed_test.sol |     -0.100% |      0.000% |      0.000% | 

</details>

bshastry commented:
> constructor/arrays_in_constructors.sol | 1.222% | 0.000% | 0.000%

Is this because the keccaks in constructor are evaluated leading to larger constants and in turn a gas increase? iiuc, this PR accounts for creation context, so the keccak optimizer would be aware if gas usage is going to increase?

hrkrshnn commented on test/cmdlineTests/keccak_optimization_deploy_code/input.sol:
```diff
@@ -0,0 +1,21 @@
+        assembly {
+            mstore(0, 100)
+            // because this is part of deploy code, the keccak will not be evaluated
+            sstore(0, keccak256(0, 32))
```
@bshastry In this example, you can see that keccak in constructors are not evaluated.

bshastry replied:
Will take a closer look at the sol test that is showing an increase

hrkrshnn commented:
After the rebase, the rule was able to affect 76 files! Instead of 13 before (https://github.com/ethereum/solidity/pull/11026#issuecomment-804000130)

Notably, the rebase contains FunctionSpecializer!


<details><summary>Click for a table of gas differences</summary>

| file name | IR-Opti | Legacy-Opti | Legacy|
|-----------|---------|-------------|-------|
| abiEncoderV1/abi_decode_v2_storage.sol |     -0.043% |      0.000% |      0.000% |
| abiEncoderV1/struct/struct_storage_ptr.sol |     -0.129% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_v2.sol |     -0.106% |      0.000% |      0.000% |
| abiEncoderV2/storage_array_encoding.sol |     -0.045% |      0.000% |      0.000% |
| abiencodedecode/abi_decode_simple_storage.sol |     -0.064% |      0.000% |      0.000% |
| array/arrays_complex_from_and_to_storage.sol |     -0.020% |      0.000% |      0.000% |
| array/byte_array_storage_layout.sol |     -0.062% |      0.000% |      0.000% |
| array/byte_array_transitional_2.sol |     -0.519% |      0.000% |      0.000% |
| array/bytes_length_member.sol |     -0.038% |      0.000% |      0.000% |
| array/dynamic_array_cleanup.sol |     -0.153% |      0.000% |      0.000% |
| array/dynamic_arrays_in_storage.sol |     -0.765% |      0.000% |      0.000% |
| array/dynamic_multi_array_cleanup.sol |     -0.651% |      0.000% |      0.000% |
| array/copying/array_copy_cleanup_uint128.sol |     -0.410% |      0.000% |      0.000% |
| array/copying/array_copy_cleanup_uint40.sol |     -0.808% |      0.000% |      0.000% |
| array/copying/array_copy_clear_storage.sol |     -0.270% |      0.000% |      0.000% |
| array/copying/array_copy_clear_storage_packed.sol |     -0.405% |      0.000% |      0.000% |
| array/copying/array_copy_different_packing.sol |     -0.294% |      0.000% |      0.000% |
| array/copying/array_copy_including_array.sol |     -0.057% |      0.000% |      0.000% |
| array/copying/array_copy_nested_array.sol |     -0.061% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_different_base.sol |     -0.196% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_struct.sol |     -0.460% |      0.000% |      0.000% |
| array/copying/array_nested_memory_to_storage.sol |     -0.094% |      0.000% |      0.000% |
| array/copying/array_of_struct_calldata_to_storage.sol |     -0.136% |      0.000% |      0.000% |
| array/copying/array_of_struct_memory_to_storage.sol |     -0.138% |      0.000% |      0.000% |
| array/copying/array_of_structs_containing_arrays_calldata_to_storage.sol |     -0.303% |      0.000% |      0.000% |
| array/copying/array_of_structs_containing_arrays_memory_to_storage.sol |     -0.165% |      0.000% |      0.000% |
| array/copying/arrays_from_and_to_storage.sol |     -0.032% |      0.000% |      0.000% |
| array/copying/bytes_storage_to_storage.sol |     -1.417% |      0.000% |      0.000% |
| array/copying/calldata_array_dynamic_to_storage.sol |     -0.074% |      0.000% |      0.000% |
| array/copying/copy_byte_array_in_struct_to_storage.sol |     -0.048% |      0.000% |      0.000% |
| array/copying/copy_byte_array_to_storage.sol |     -0.114% |      0.000% |      0.000% |
| array/copying/copy_removes_bytes_data.sol |     -0.024% |      0.000% |      0.000% |
| array/copying/memory_dyn_2d_bytes_to_storage.sol |     -0.604% |      0.000% |      0.000% |
| array/copying/storage_memory_nested.sol |     -0.244% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_bytes.sol |     -0.059% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_from_pointer.sol |     -0.244% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_struct.sol |     -0.189% |      0.000% |      0.000% |
| array/copying/storage_memory_packed_dyn.sol |     -0.816% |      0.000% |      0.000% |
| array/copying/array_copy_calldata_storage.sol |     -0.014% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_dynamic_dynamic.sol |     -0.166% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_static_dynamic.sol |     -0.090% |      0.000% |      0.000% |
| array/copying/array_copy_storage_to_memory_nested.sol |     -2.262% |      0.000% |      0.000% |
| array/copying/array_nested_calldata_to_storage.sol |     -0.123% |      0.000% |      0.000% |
| array/copying/copy_function_storage_array.sol |     -0.154% |      0.000% |      0.000% |
| array/delete/bytes_delete_element.sol |     -0.894% |      0.000% |      0.000% |
| array/delete/delete_storage_array_packed.sol |     -0.289% |      0.000% |      0.000% |
| array/pop/array_pop_array_transition.sol |     -0.262% |      0.000% |      0.000% |
| array/pop/array_pop_uint16_transition.sol |     -0.819% |      0.000% |      0.000% |
| array/pop/array_pop_uint24_transition.sol |     -0.737% |      0.000% |      0.000% |
| array/pop/byte_array_pop_copy_long.sol |     -0.101% |      0.000% |      0.000% |
| array/pop/byte_array_pop_long_storage_empty.sol |     -0.287% |      0.000% |      0.000% |
| array/pop/byte_array_pop_long_storage_empty_garbage_ref.sol |     -0.289% |      0.000% |      0.000% |
| array/pop/byte_array_pop_masking_long.sol |     -0.134% |      0.000% |      0.000% |
| array/push/array_push_nested_from_calldata.sol |     -0.134% |      0.000% |      0.000% |
| array/push/array_push_packed_array.sol |     -0.380% |      0.000% |      0.000% |
| array/push/array_push_struct.sol |     -0.137% |      0.000% |      0.000% |
| array/push/array_push_struct_from_calldata.sol |     -0.135% |      0.000% |      0.000% |
| array/push/byte_array_push_transition.sol |     -0.610% |      0.000% |      0.000% |
| array/push/push_no_args_2d.sol |     -0.016% |      0.000% |      0.000% |
| array/push/push_no_args_bytes.sol |     -0.409% |      0.000% |      0.000% |
| array/push/array_push.sol |     -0.222% |      0.000% |      0.000% |
| constructor/bytes_in_constructors_packer.sol |      1.679% |      0.000% |      0.000% |
| constructor/arrays_in_constructors.sol |      1.198% |      0.000% |      0.000% |
| structs/struct_containing_bytes_copy_and_delete.sol |     -0.031% |      0.000% |      0.000% |
| structs/struct_delete_storage_nested_small.sol |     -0.205% |      0.000% |      0.000% |
| structs/struct_delete_storage_with_array.sol |     -0.230% |      0.000% |      0.000% |
| structs/struct_delete_storage_with_arrays_small.sol |     -0.113% |      0.000% |      0.000% |
| structs/calldata/calldata_struct_with_nested_array_to_storage.sol |     -0.085% |      0.000% |      0.000% |
| various/skip_dynamic_types_for_structs.sol |     -0.026% |      0.000% |      0.000% |
| various/destructuring_assignment.sol |     -0.213% |      0.000% |      0.000% |
| viaYul/array_storage_index_access.sol |     -0.649% |      0.000% |      0.000% |
| viaYul/array_storage_index_boundary_test.sol |     -1.420% |      0.000% |      0.000% |
| viaYul/array_storage_push_empty.sol |     -1.676% |      0.000% |      0.000% |
| viaYul/array_storage_push_empty_length_address.sol |     -1.457% |      0.000% |      0.000% |
| viaYul/array_storage_push_pop.sol |     -1.125% |      0.000% |      0.000% |
| viaYul/array_storage_index_zeroed_test.sol |     -0.794% |      0.000% |      0.000% |

</details>

chriseth commented:
Ah, that's really nice! Do you have any values on the code size increase for real-world contracts?

hrkrshnn commented:
@chriseth I don't have gas data from any real world contract available. I can check this.

Also, it seems that deploy costs are not mentioned if there is no `constructor()` call in isoltest gas update. Let me see if it's possible to improve that. 

Anyway, contracts can disable this by specifying an `--optimize-run` value below `148`.

chriseth commented:
We should really be careful about this. It could mean that event signatures are always included as 32-byte constants, which is extremely expensive and can bring contracts closer to the code size limit.

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+	// Deploy time increase
+	// [PUSH32 32-byte-value]: 33 bytes
+	// instead of
+	// [PUSH1 0 PUSH1 32 KECCAK256]: 5 bytes
```
I think it is plausible to assume that 0 and 32 are available somewhere on the stack, so these are rather DUPs than PUSHes

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
+	std::vector<Expression> const& _arguments
+)
+{
+	// Runtime saving: [PUSH32 32-byte-value] instead of [PUSH1 0 PUSH1 32 KECCAK256]
```
Can you use GasMeter in EVMMetrics.h instead? Maybe you can also add a function `literalGas(size_t _bytes)` that you then call from `GasMeterVisitor::operator()(Literal const& _lit)`?

hrkrshnn committed:
GasMeter can now account for gas of Keccak-256

hrkrshnn committed:
Isoltest: updated the gas tests.

hrkrshnn commented:
> It could mean that event signatures are always included as 32-byte constants

@chriseth Event signatures are already computed as 32 byte constants in IR. This PR doesn't seem to affect contracts like ERC20.

chriseth commented on libsolidity/codegen/CompilerContext.cpp:
```diff
@@ -548,6 +548,7 @@ void CompilerContext::optimizeYul(yul::Object& _object, yul::EVMDialect const& _
 		_object,
 		_optimiserSettings.optimizeStackAllocation,
 		_optimiserSettings.yulOptimiserSteps,
+		isCreation? nullopt : make_optional(_optimiserSettings.expectedExecutionsPerDeployment),
```
Why nullopt instead of 1?

hrkrshnn replied:
I think it's not enough to treat creation context as having `expectedExecutionsPerDeployment` as 1. This is because in creation context data gas is cheaper than at runtime context.

chriseth commented on libyul/backends/evm/EVMMetrics.cpp:
```diff
@@ -117,6 +117,9 @@ void GasMeterVisitor::instructionCostsInternal(evmasm::Instruction _instruction)
 	if (_instruction == evmasm::Instruction::EXP)
 		m_runGas += evmasm::GasCosts::expGas + evmasm::GasCosts::expByteGas(m_dialect.evmVersion());
+	else if (_instruction == evmasm::Instruction::KECCAK256)
+		// only account for Keccak256 on a single word
```
Please document this in the class description in the header file.

chriseth commented on libyul/backends/wasm/EVMToEwasmTranslator.cpp:
```diff
@@ -68,7 +68,8 @@ Object EVMToEwasmTranslator::run(Object const& _object)
 	set<YulString> reservedIdentifiers;
 	NameDispenser nameDispenser{m_dialect, ast, reservedIdentifiers};
-	OptimiserStepContext context{m_dialect, nameDispenser, reservedIdentifiers};
+	// 200 is the default value for optimize-runs
```
Please document that it is unused - that is the point here, right?
In addition, can you use `OptimiserSettings::standard().expectedExecutinosPerDeployment`?

chriseth commented on libyul/optimiser/DataFlowAnalyzer.h:
```diff
@@ -134,6 +136,9 @@ class DataFlowAnalyzer: public ASTModifier
 	/// Returns true iff the variable is in scope.
 	bool inScope(YulString _variableName) const;
 
+	/// Returns the literal value of the identifier, if it exists.
```
Is this useful to have in `DataflowAnalyzer`?

hrkrshnn replied:
It was initially a lambda inside LoadResolver, but @ekpyron suggested that it's useful to move it inside DataFlowAnalyzer. see: https://github.com/ethereum/solidity/pull/11026#discussion_r594218192

chriseth replied:
I would say that especially the data flow analyzer should be kept as simple as possible, but maybe it's fine to put this function here since it deals with a member of the class.

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -38,21 +49,39 @@ void LoadResolver::run(OptimiserStepContext& _context, Block& _ast)
 {
 	DataFlowAnalyzer::visit(_e);
 
+	auto isKeccak = [&](YulString const& _name) -> bool
```
Can you use the same mechanism that we are already using for mload, i.e. define a function in the dialect and then compare by name?

hrkrshnn replied:
Sure, I'll implement this.

hrkrshnn replied:
I actually implemented this. But I think it ends up being an overkill. I would just prefer the lambda.

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,61 @@ void LoadResolver::tryResolve(
+		m_expectedExecutionsPerDeployment ? *m_expectedExecutionsPerDeployment : 1
+	};
+
+	auto costOfKeccak = gasMeter.costs(_e);
```
I'm sorry but I really want to know if this is a `u256` or a `uint64_t`.

chriseth replied:
Ok, I looked it up and it is a `size_t`. I'm wondering: Is that value large enough? Shouldn't we change `GasMeter::costs` to return a u256 instead?

hrkrshnn replied:
Making it u256 makes sense. However, EIP-1985 proposes an upper bound of `2**63 - 1` for `gas()`, still it `u256` would be a safe option for us.

Anyway, this change should be a separate PR.

chriseth replied:
It's unrelated to `gas()`: The issue here is that this value is the product of the gas costs and the runs parameter.

chriseth replied:
I acutally think it should be a `bigint` just as the numbers returned by the other gas meter.

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,61 @@ void LoadResolver::tryResolve(
+			{},
+			LiteralKind::Number,
+			// a dummy 256-bit number to represent the Keccak256 hash.
+			YulString{u256(-1).str()},
```
`numeric_limits<256>::max()`?

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,61 @@ void LoadResolver::tryResolve(
+		!m_expectedExecutionsPerDeployment,
+		m_expectedExecutionsPerDeployment ? *m_expectedExecutionsPerDeployment : 1
+	};
+
```
Please add a comment: The costs are only correct for keccak hashes of up to 32 bytes.

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,61 @@ void LoadResolver::tryResolve(
+
+	yulAssert(_arguments.size() == 2, "");
+	Identifier const* memoryKey = std::get_if<Identifier>(&_arguments.at(0));
+	Identifier const* offset = std::get_if<Identifier>(&_arguments.at(1));
```
Isn't this `length`?

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -76,3 +105,61 @@ void LoadResolver::tryResolve(
+	if (!memoryKey || !offset)
+		return;
+
+	if (
+		auto memoryValue = util::valueOrNullptr(m_memory, memoryKey->name);
```
```suggestion
	YulString const* memoryValue = util::valueOrNullptr(m_memory, memoryKey->name);
	if (
```

chriseth commented:
Can you add an inline assembly test that uses `keccak(0, 16)`?

hrkrshnn commented:
I also realize that we have to skip the step if `msize` is in the AST. I'll fix this.

hrkrshnn committed:
Skip resolving Keccak if AST contains msize.

Also a refactoring changing `m_optimizeMLoad` to `m_containsMSize`.

hrkrshnn committed:
Tests and changelog

hrkrshnn commented:
Rebased this to fix a conflict.

chriseth commented on test/tools/yulopti.cpp:
```diff
@@ -191,7 +191,7 @@ class YulOpti
 			cout << ' ' << option << endl;
 
-			OptimiserStepContext context{m_dialect, *m_nameDispenser, reservedIdentifiers};
+			OptimiserStepContext context{m_dialect, *m_nameDispenser, reservedIdentifiers, 200};
```
Can't we get this number from somewhere? Same below.

chriseth commented on test/libyul/YulOptimizerTestCommon.cpp:
```diff
@@ -434,6 +434,7 @@ void YulOptimizerTestCommon::updateContext()
 		*m_nameDispenser,
-		m_reservedIdentifiers
+		m_reservedIdentifiers,
+		200 // default value of optimize-runs
```
Again, can we get this number from somewhere? Then we don't need a comment.

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -71,8 +102,64 @@ void LoadResolver::tryResolve(
+		return;
+
+	auto memoryValue = util::valueOrNullptr(m_memory, memoryKey->name);
+	if (memoryValue && inScope(*memoryValue))
```
Do you need the `inScope` check?

hrkrshnn replied:
I think it's redundant, but it was added during `DataFlowAnalyzer` refactoring (https://github.com/ethereum/solidity/pull/10467/commits/75bed10be56939d7b4a28bfc9fa006f4fe1267d9) to make `LoadResolver` slightly safer.

chriseth replied:
Ok, but it's not relevant here, because we don't use it directly, is it?

hrkrshnn replied:
We still get the value to hash from `memoryValue`. So I think the check makes it slightly safer, although redundant.

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -71,8 +102,64 @@ void LoadResolver::tryResolve(
+		if (memoryContent && byteLength && *byteLength <= 32)
+		{
+			bytes contentAsBytes = toBigEndian(*memoryContent);
+			bytes slicedBytes = contentAsBytes | ranges::views::take(unsigned(*byteLength)) | ranges::to<vector>;
```
```suggestion
			contentAsBytes.resize(static_cast<size_t>(*byteLength));
```

hrkrshnn committed:
Added hashFunction to Dialect. For EVMDialect, it is keccak256

In a later PR, the functions `storageLoadFunction`, `storageStoreFunctions`, etc will be refactored
to return a YulString instead of a pointer to a BuiltinFunction.

hrkrshnn committed:
Changed the type of gas calculation to bigint instead of size_t

Since the gas calculation can involve multiplication by ``--optimize-runs``, it is possible that
`size_t` is not enough to represent the total gas.

chriseth commented on libyul/backends/evm/ConstantOptimiser.h:
```diff
@@ -58,7 +58,7 @@ class ConstantOptimiser: public ASTModifier
 	{
 		std::unique_ptr<Expression> expression;
-		size_t cost = size_t(-1);
+		bigint cost = size_t(-1);
```
What does this mean?

hrkrshnn replied:
I guess treating `2**64 - 1` as infinity.

chriseth replied:
Sure, but is this mechanism still working?

hrkrshnn replied:
I actually think the initial value is never used anywhere. I'll leave it as the old one.

chriseth replied:
but shouldn't it rather be zero then?

hrkrshnn replied:
Updated it to 0

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -71,8 +87,64 @@ void LoadResolver::tryResolve(
+		m_expectedExecutionsPerDeployment ? *m_expectedExecutionsPerDeployment : 1
+	};
+
+	auto costOfKeccak = gasMeter.costs(_e);
```
```suggestion
	bigint costOfKeccak = gasMeter.costs(_e);
```

chriseth commented on libyul/optimiser/LoadResolver.cpp:
```diff
@@ -71,8 +87,64 @@ void LoadResolver::tryResolve(
+	};
+
+	auto costOfKeccak = gasMeter.costs(_e);
+	auto costOfLiteral = gasMeter.costs(
```
```suggestion
	bigint costOfLiteral = gasMeter.costs(
```

chriseth commented on libyul/backends/evm/EVMMetrics.h:
```diff
@@ -36,40 +36,42 @@ struct EVMDialect;
+	GasMeter(EVMDialect const& _dialect, bool _isCreation, bigint _runs):
 		m_dialect(_dialect),
 		m_isCreation{_isCreation},
 		m_runs(_isCreation? 1 : _runs)
```
move?

chriseth reviewed (approved):
Could add `move` and maybe also squash some commits, but fine otherwise.
